sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTranslation.java

```diff
@@ -184,105 +189,124 @@ public Row toConfigRow(TypedRead<?> transform) {
-      }
-      String kmsKey = configRow.getString("kms_key");
-      if (kmsKey != null) {
-        builder = builder.setKmsKey(kmsKey);
-      }
-      byte[] typeDescriptorBytes = configRow.getBytes("type_descriptor");
-      if (typeDescriptorBytes != null) {
-        builder = builder.setTypeDescriptor((TypeDescriptor) fromByteArray(typeDescriptorBytes));
-      }
-      byte[] toBeamRowFnBytes = configRow.getBytes("to_beam_row_fn");
-      if (toBeamRowFnBytes != null) {
-        builder = builder.setToBeamRowFn((ToBeamRowFunction) fromByteArray(toBeamRowFnBytes));
-      }
-      byte[] fromBeamRowFnBytes = configRow.getBytes("from_beam_row_fn");
-      if (fromBeamRowFnBytes != null) {
-        builder = builder.setFromBeamRowFn((FromBeamRowFunction) fromByteArray(fromBeamRowFnBytes));
-      }
-      Boolean useAvroLogicalTypes = configRow.getBoolean("use_avro_logical_types");
-      if (useAvroLogicalTypes != null) {
-        builder = builder.setUseAvroLogicalTypes(useAvroLogicalTypes);
-      }
-      Boolean projectionPushdownApplied = configRow.getBoolean("projection_pushdown_applied");
-      if (projectionPushdownApplied != null) {
-        builder = builder.setProjectionPushdownApplied(projectionPushdownApplied);
-      }
+        String jsonTableRef = configRow.getString("json_table_ref");
+        if (jsonTableRef != null) {
+          builder = builder.setJsonTableRef(StaticValueProvider.of(jsonTableRef));
+        }
+        String query = configRow.getString("query");
+        if (query != null) {
+          builder = builder.setQuery(StaticValueProvider.of(query));
+        }
+        Boolean validate = configRow.getBoolean("validate");
+        if (validate != null) {
+          builder = builder.setValidate(validate);
+        }
+        Boolean flattenResults = configRow.getBoolean("flatten_results");
+        if (flattenResults != null) {
+          builder = builder.setFlattenResults(flattenResults);
+        }
+        Boolean useLegacySQL = configRow.getBoolean("use_legacy_sql");
+        if (useLegacySQL != null) {
+          builder = builder.setUseLegacySql(useLegacySQL);
+        }
+        Boolean withTemplateCompatibility = configRow.getBoolean("with_template_compatibility");
+        if (withTemplateCompatibility != null) {
+          builder = builder.setWithTemplateCompatibility(withTemplateCompatibility);
+        }
+        byte[] bigqueryServicesBytes = configRow.getBytes("bigquery_services");
+        if (bigqueryServicesBytes != null) {
+          try {
+            builder =
+                builder.setBigQueryServices(
+                    (BigQueryServices) fromByteArray(bigqueryServicesBytes));
+          } catch (InvalidClassException e) {
+            LOG.warn(
+                "Could not use the provided `BigQueryServices` implementation when upgrading."
+                    + "Using the default.");
+            builder.setBigQueryServices(new BigQueryServicesImpl());
```

Comment:
If this is sufficient for correct behavior, why do we deserialize the service bytes at all?